contracts/token/ERC721/extensions/draft-ERC721Royalty.sol

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts v4.4.0 (token/ERC721/extensions/draft-ERC721Royalty.sol)
+
+pragma solidity ^0.8.0;
+
+import "../ERC721.sol";
+import "./draft-IERC721Royalty.sol";
+import "../../../utils/introspection/ERC165.sol";
+
+/**
+ * @dev Implementation of the ERC721 Royalty extension allowing royalty information to be stored and retrieved, as defined in
+ * https://eips.ethereum.org/EIPS/eip-2981[EIP-2981].
+ *
+ * Adds the {_setTokenRoyalty} methods to set the token royalty information, and {_setGlobalRoyalty} method to set a global
+ * royalty information.
+ *
+ * _Available since v4.5._
+ */
+abstract contract ERC721Royalty is IERC721Royalty, ERC721 {
+    struct RoyaltyInfo {
+        address receiver;
+        uint256 royaltyFraction;
+    }
+
+    RoyaltyInfo private _globalRoyaltyInfo;
+    mapping(uint256 => RoyaltyInfo) private _tokenRoyaltyInfo;
+
+    /**
+     * @dev Sets tokens royalties
+     *
+     * Requirements:
+     * - `tokenId` must be already mined.
+     * - `receiver` cannot be the zero address.
+     * - `fraction` must indicate the percentage fraction using two decimals.
+     */
+    function _setTokenRoyalty(
+        uint256 tokenId,
+        address receiver,
+        uint256 fraction
+    ) internal virtual {
+        require(fraction > 0, "ERC2981: Royalty percentage is too low");
+        require(fraction <= _feeDenominator(), "ERC2981: Royalty percentage will exceed salePrice");
+        require(receiver != address(0), "ERC2981: Invalid receiver");
+        require(_exists(tokenId), "ERC2981: Nonexistent token");
+
+        _tokenRoyaltyInfo[tokenId] = RoyaltyInfo(receiver, fraction);
+    }
+
+    /**
+     *
+     * @dev Sets global royalty
+     *
+     * Requirements:
+     * - `receiver` cannot be the zero address.
+     * - `fraction` must indicate the percentage fraction. Needs to be set appropriately
+     * according to the _feeDenominator granularity.
+     */
+    function _setGlobalRoyalty(address receiver, uint256 fraction) internal virtual {
+        require(fraction > 0, "ERC2981: Royalty percentage is too low");
+        require(fraction <= _feeDenominator(), "ERC2981: Royalty percentage will exceed salePrice");
+        require(receiver != address(0), "ERC2981: Invalid receiver");
+
+        _globalRoyaltyInfo = RoyaltyInfo(receiver, fraction);
+    }
+
+    /**
+     * @dev See {IERC721Royalty-royaltyInfo}
+     */
+    function royaltyInfo(uint256 _tokenId, uint256 _salePrice) external view override returns (address, uint256) {
+        RoyaltyInfo memory royalty;
+
+        if (_tokenRoyaltyInfo[_tokenId].receiver != address(0)) {
```

Comment:
if we change the `royaltyFraction` to be a uint96, then `RoyaltyInfo` would be a single slot ... which could then be optimized to avoid doing the ssload twice:

```
RoyaltyInfo memory royalty = _tokenRoyaltyInfo[_tokenId];
if (royalty.receiver == address(0)) {
    royalty = _globalRoyaltyInfo;
}

...
```